File: Cpp/BaekJoon/Gold/G5_1068.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> tree;
vector<bool> check;

int n, skip;
int leapCount = 0;

void checkLeap(int now) {
  if (check[now] || now == skip) return;

  check[now] = true;

  bool hasChild = false;

  for (int i = 0; i < tree[now].size(); i++) {
    int next = tree[now][i];

    if (next == skip) continue;

    hasChild = true;
    checkLeap(next);
  }

  if (!hasChild) leapCount++;

  return;
}

int main() {
  cin >> n;

  tree.resize(n);
  check.resize(n);

  int root = 0;

  for (int i = 0; i < n; i++) {
    int parent;
    cin >> parent;

    if (parent == -1) {
      root = i;
      continue;
    }

    tree[parent].push_back(i);
  }

  cin >> skip;

  checkLeap(root);

  cout << leapCount;

  return 0;
}

File: Java/BaekJoon/Gold/G5_1068.java
package BaekJoon.Gold;

import java.util.ArrayList;
import java.util.Scanner;

public class G5_1068 {
    static int n, skipNode;
    static int leapCount = 0;

    static ArrayList<ArrayList<Integer>> tree = new ArrayList<>();

    public static void checkLeap(int now) {
        if (now == skipNode) {
            return;
        }

        boolean hasChild = false;

        for (int i = 0; i < tree.get(now).size(); i++) {
            int next = tree.get(now).get(i);

            if (next == skipNode) {
                continue;
            }

            hasChild = true;
            checkLeap(next);
        }

        if (!hasChild) {
            leapCount++;
        }

        return;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();

        // 이중리스트 생성
        for (int i = 0; i < n; i++) {
            tree.add(new ArrayList<>());
        }

        int root = -1;

        // 부모노드 입력
        for (int i = 0; i < n; i++) {
            int parent = sc.nextInt();

            if (parent == -1) {
                root = i;
                continue;
            }

            tree.get(parent).add(i);
        }

        // 제외 노드 입력
        skipNode = sc.nextInt();

        // 리프 노드 탐색
        checkLeap(root);

        // 결과 출력
        System.out.println(leapCount);

        sc.close();
    }
}
